java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java

```diff
@@ -115,49 +130,40 @@ public void computedFrom(List<SymbolicValue> symbolicValues) {
   }
 
   public List<ProgramState> setConstraint(ProgramState programState, ObjectConstraint nullConstraint) {
-    Object data = programState.getConstraint(this);
-    if (data instanceof ObjectConstraint) {
-      ObjectConstraint nc = (ObjectConstraint) data;
-      if (nc.isNull() ^ nullConstraint.isNull()) {
-        // setting null where value is known to be non null or the contrary
-        return ImmutableList.of();
-      } else if (hasAnyStatus(nc) && hasNoStatus(nullConstraint)) {
-        return ImmutableList.of(programState);
+    Constraint cstraint = programState.getConstraint(this, nullConstraint.getClass());
+    if (cstraint == null) {
+      if(nullConstraint.isNull()) {
+        PMap<Class<? extends Constraint>, Constraint> onlyNullConstraint = PCollections.<Class<? extends Constraint>, Constraint>emptyMap()
+          .put(ObjectConstraint.class, ObjectConstraint.NULL);
+        return ImmutableList.of(programState.addConstraints(this, onlyNullConstraint));
```

Comment:
I would add a comment saying that setting null erases all other known constraints.